### PR TITLE
Extend and improve RecursiveApply, fix DSS inference

### PR DESCRIPTION
**Pull request by charleskawczynski**
This PR:
 - Extends and improves RecursiveApply (incorporates some changes/functionality in https://github.com/CliMA/ClimaCore.jl/pull/1326) by working with Tuples/NamedTuples in a type-stable way
 - Adds RecursiveApply tests
 - Fix DSS inference and corresponding test

### CODE REVIEW

**Comment by charleskawczynski** [comment]
bors try

File: src/RecursiveApply/RecursiveApply.jl [hunk 1]
```diff
@@ -9,19 +9,51 @@ module RecursiveApply
...
+
+rmaptype_Tuple(_, ::Type{Tuple{}}, ::Type{T}) where {T <: Tuple} = ()
+rmaptype_Tuple(_, ::Type{T}, ::Type{Tuple{}}) where {T <: Tuple} = ()
+rmaptype_Tuple(
+    fn::F,
+    ::Type{T1},
+    ::Type{T2},
+) where {F, T1 <: Tuple, T2 <: Tuple} = (
+    rmaptype(fn, first_param(T1), first_param(T2)),
+    rmaptype_Tuple(rmaptype, fn, tail_params(T1), tail_params(T2))...,
```
Comment (dennisYatunin) [inline src/RecursiveApply/RecursiveApply.jl]:
```suggestion
    rmaptype_Tuple(fn, tail_params(T1), tail_params(T2))...,
```

**Review by dennisYatunin — APPROVED** [review]

**Comment by charleskawczynski** [comment]
Ambiguities seem to have increased from dependency patch updates (none are from RecursiveApply)

**Pull request merged**